internal/policy/verify.go

```diff
@@ -0,0 +1,140 @@
+	entryStack := []*rsl.Entry{lastEntry}
+
+	var currentPolicy *State
+	// 1. Identify policy applicable at firstEntry
+	policyEntry, err := rsl.GetLatestEntryForRefBefore(repo, PolicyRef, firstEntry.ID)
+	if err != nil {
+		return err
+	}
+	state, err := LoadStateForEntry(ctx, repo, policyEntry)
+	if err != nil {
+		return err
+	}
+	currentPolicy = state
+
+	// 2. Enumerate RSL entries between firstEntry and lastEntry, ignoring irrelevant ones
+	iteratorEntry := lastEntry
+	for {
+		if iteratorEntry.GetID() == firstEntry.ID {
+			break
+		}
+
+		parentEntryTmp, err := rsl.GetParentForEntry(repo, iteratorEntry)
+		if err != nil {
+			return err
+		}
+		parentEntry := parentEntryTmp.(*rsl.Entry) // TODO: handle annotations
+
+		if parentEntry.RefName == target || parentEntry.RefName == PolicyRef {
+			entryStack = append(entryStack, parentEntry)
+		}
+
+		iteratorEntry = parentEntry
+	}
+
+	// entryStack has a list of RSL entries in reverse order
+	entryQueue := []*rsl.Entry{}
+	for j := len(entryStack) - 1; j >= 0; j-- {
+		// TODO: instead of reversing it, we can likely just protest this in
+		// reverse order but the policy states are the issue
+		entryQueue = append(entryQueue, entryStack[j])
+	}
+
+	for _, entry := range entryQueue {
+		// FIXME: we're not verifying policy RSL entry signatures because
+		// we need to establish how to fetch that info. There's some code right
+		// now in State that sets up all keys listed in policies as applicable
+		// signers for that state but it's unclear if that's the way to go. An
+		// additional blocker is for managing special keys like root and targets
+		// keys. RSL entry signatures are commit signatures. What do we do when
+		// metadata is signed using other methods? Do we instead flip the script
+		// and require metadata signatures to match git signing methods?
+		if entry.RefName == PolicyRef {
+			// TODO: this is repetition if the firstEntry is for policy
+			state, err := LoadStateForEntry(ctx, repo, entry)
+			if err != nil {
+				return err
+			}
+			currentPolicy = state
+
+			continue
+		}
+
+		if err := verifyEntry(ctx, repo, currentPolicy, entry); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func verifyEntry(ctx context.Context, repo *git.Repository, policy *State, entry *rsl.Entry) error {
+	var (
+		trustedKeys []*tuf.Key
+		err         error
+	)
+
+	// 1. Find authorized public keys for entry's ref
+	if entry.RefName == PolicyRef {
+		trustedKeys = policy.AllPublicKeys
+	} else {
+		trustedKeys, err = policy.FindPublicKeysForPath(ctx, fmt.Sprintf("git:%s", entry.RefName)) // FIXME: "git:" shouldn't be here
+		if err != nil {
+			return err
+		}
+	}
+
+	commitObj, err := repo.CommitObject(entry.ID)
+	if err != nil {
+		return err
+	}
+
+	for _, key := range trustedKeys {
+		if err := gitinterface.VerifyCommitSignature(commitObj, key); err != nil {
+			if !errors.Is(err, gitinterface.ErrIncorrectVerificationKey) {
+				return err
+			}
+		} else {
+			return nil
+		}
```

Comment:
This is probably a case where you may want to reverse the check to make it clearer under what conditions the loop will continue.

```suggestion
		err := gitinterface.VerifyCommitSignature(commitObj, key)
		if err == nil {
			// Key is valid
			return nil
		}
		if !errors.Is(err, gitinterface.ErrIncorrectVerificationKey) {
			return err
		}
		// Haven't found a valid key - continue
```

Reply:
Much cleaner, thanks!